src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -812,6 +816,646 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
+             "load from cache doesn't access the cache array?");
+      intptr_t array_cache_load_offset_in_body = array_cache_load_const_offset - header_size_for_cache_array_load;
+
+      Node* index_in_cache_array = _kit.gvn().intcon(
+              checked_cast<int>(array_cache_load_offset_in_body >> shift_for_cache_array_load));
+      if (second_addp_for_array_cache_load_adr->is_AddP()) {
+        assert(!second_addp_for_array_cache_load_adr->in(AddPNode::Address)->is_AddP() &&
+               second_addp_for_array_cache_load_adr->in(AddPNode::Base) == array_cache_load_adr->in(AddPNode::Base),
+               "only 2 AddPs for address computation");
+        Node* array_cache_load_offset_from_second_addp = second_addp_for_array_cache_load_adr->in(AddPNode::Offset);
+        assert(array_cache_load_offset_from_second_addp->Opcode() == Op_LShiftX &&
+               array_cache_load_offset_from_second_addp->in(2)->find_int_con(-1) == shift_for_cache_array_load,
+               "Not an array access?");
+        Node* array_cache_load_index_from_second_addp = array_cache_load_offset_from_second_addp->in(1);
+#ifdef _LP64
+        assert(array_cache_load_index_from_second_addp->Opcode() == Op_ConvI2L,
+               "unexpected address calculation shape");
+        array_cache_load_index_from_second_addp = array_cache_load_index_from_second_addp->in(1);
+        assert(!(array_cache_load_index_from_second_addp->Opcode() == Op_CastII &&
+                 array_cache_load_index_from_second_addp->in(0)->is_Proj() &&
+                 array_cache_load_index_from_second_addp->in(0)->in(0) == _cache_not_null_iff),
+               "no CastII because index_in_cache_array is known to be positive");
+#endif
+        index_in_cache_array = _kit.gvn().transform(new AddINode(array_cache_load_index_from_second_addp, index_in_cache_array));
+      }
+
+      if (_first_cache_probe_iff == nullptr) {
+        _first_cache_probe_iff = maybe_iff->as_If();
+        _first_index_in_cache = index_in_cache_array;
+      } else {
+        assert(_second_cache_probe_iff == nullptr, "no more than 2 cache probes");
+        _second_cache_probe_iff = maybe_iff->as_If();
+        _second_index_in_cache = index_in_cache_array;
+      }
+      return true;
+    }
+
+    // First traversal of the get() subgraph starts from the end of the method and follows control paths until it reaches
+    // the Thread.scopedValueCache() call. Given the shape of the method and some paths may have been trimmed and end with
+    // an uncommon trap, it could reach either the first or the second cache probe if first. Figure out which is the first
+    // here.
+    void adjust_order_or_first_and_second_probe_if(const Unique_Node_List &scoped_value_get_subgraph) {
+      if (_second_cache_probe_iff == nullptr) {
+        return;
+      }
+      assert(_first_cache_probe_iff != nullptr, "can't have a second iff if there's no first one");
+      Node_Stack stack(0);
```

Comment:
```suggestion
      ResourceMark rm;
      Node_Stack stack(0);
```
Would that work?

Reply:
Plus: it still makes me a bit nervous that there is not visited set... But I guess if we needed one, then we would have already walked outside of the relevant subgraph, and things would be broken... hmm